packages/xstate-react/src/useMachine.ts

```diff
@@ -81,21 +76,36 @@ export function useMachine<
         nextState.changed === undefined &&
         Object.keys(nextState.children).length;
 
-      if (nextState.changed || initialStateChanged) {
-        setState(nextState);
-      }
+      return !(nextState.changed || initialStateChanged);
+    },
+    [service]
+  );
+
+  const subscribe = useCallback(
+    (handleStoreChange) => {
+      const { unsubscribe } = service.subscribe(handleStoreChange);
+      return unsubscribe;
     },
-    []
+    [service]
+  );
+  const storeSnapshot = useSyncExternalStoreWithSelector(
+    subscribe,
+    getSnapshot,
+    undefined,
+    identity,
+    isEqual
   );
 
-  const service = useInterpret(getMachine, options, listener);
+  useEffect(() => {
+    const rehydratedState = options.state;
+    service.start(
```

Comment:
After those changes, `useMachine` no longer depends on the `useInterpret`. `useInterpret` starts the machine in the layout effect but this one has to do that in the passive effect. Maybe we should change the behavior of the `useInterpret` too - but I've wanted to avoid changing too much now. I've also wanted to change this for now because how the listener/subscriber gets wired is now different and has to be managed here, in the `useMachine`, but it was previously managed by `useInterpret`. Technically, we could just not pass any subscriber to the `useInterpret` and just subscribe with `useSyncExternalStore` here... and maybe that is something that we should do.

Reply:
Actually, I've added a comment on why we can't use `useInterpret` here right now:
https://github.com/statelyai/xstate/pull/2939/commits/47f566cc83e06bac152cfd599a4bd6c786f1183b#diff-19f47c3f58ef51d8a79a32343f5c32c02d6d7237ecc29f636cb53846d00424d5R68-R69